javascript-review.js

```diff
@@ -4,112 +4,273 @@
 
+function Student(name) {
+    this.name = name;
+    this.favColor = red;
+}
+
+
+
+
+
 // Note that a constructor function is the old school syntax that was used in javascript before the sugar class syntax came along. 
 
+
+
+
 // 3.
 
 // Can you access variables defined outside of functions within the scope of functions in JS? Show a simple example of this being used? (Define a constant in global scope and use that in a very simple function.)
 
+let gVar = 5;
+
+function printgVar() {
+    console.log(gVar);
+}
+
+
+
+
 // 4.
 
 // What two ways can be used to access the properties of objects in JS? Give examples of both (although assume the objects have been defined).
 
+// use dot notation such as student1.name to access the value of the property 
+// or use the bracket notation such as student1["name"] to access the value of the property
+
+
+
 // 5.
 
 // What is JSON? Be clear but concise.
 
+// JSON is a data storage format which uses JS object notation. JSON itself is a string and it can be easily converted into JS object to facilitate CRUD data manipulation.
+
+
+
+
 // 6.
 
 // What is a callback?
 
+//callback is a lower order function nested within a higher order function. It can be passed into the higher order function without invocation. Later, some code in the higher order function can trigger the invocation of the callback.
+
+
+
 // 7.
 
 // Define a function with that takes two number arguments and a callback. The function will add the two numbers and pass them to the callback as an argument. Now call this function and in the callback simply console.log out the argument that has been passed through.
 
+function callback(param) {
+    console.log(param);
+}
+
+function takeNumsAndCb (num_param1, num_param2, callback) {
+    let res = num_param1 + num_param2;
+    callback(res); 
+} 
+
+
+
+
+
+
+
 // 8.
 
 // Why do we use promises?
 
+// When we use promise, what we want is basically to run some asynchorounous operations which are usually associated with runnning a piece of code that takes a long run to get a returned value such as getting the data from most of the WEB API . Using asynchorounous operation can make our program more responsive and faster, by allowing user to run other piece of coding while letting the async code to the background (aka event loop) 
+
+
 // 9.
 
 // What are the two methods at our disposal if we have hold of a promise? What are the names of the functions that are related to these functions (that are called when the promise is fulfilled or fails).
 
+// first method is to use .then method
+// the second method is to use aync/await 
+
+
+
+
 // 10.
 
 // What is the DOM?
 
+//It is a model to break down html elements into  objects which you can interact with and manipulate with using normal javascript code or other programming language. A key feature of this DOM structure is that it is based on a tree structure just like any complex js objects 
```

Comment:
now just html elements, it's anything in the window of the browser, we even have access to cookies